File: tools/release_helper/metadata.py
"""
App metadata utilities for the release helper.
"""

import json
from pathlib import Path
from typing import Dict, List

from tools.release_helper.core import run_bazel, find_workspace_root

# In-process cache for app metadata to avoid redundant Bazel builds
_metadata_cache: Dict[str, Dict] = {}


def _read_metadata_file(bazel_target: str) -> Dict:
    """Read a metadata JSON file from bazel-bin without building.
    
    Args:
        bazel_target: Full bazel target path (e.g., "//path/to/app:app_metadata")
        
    Returns:
        Parsed metadata dict, or None if file doesn't exist
    """
    if not bazel_target.startswith("//"):
        raise ValueError(f"Invalid bazel target format: {bazel_target}")
    
    target_parts = bazel_target[2:].split(":")
    if len(target_parts) != 2:
        raise ValueError(f"Invalid bazel target format: {bazel_target}")
    
    package_path = target_parts[0]
    target_name = target_parts[1]

    workspace_root = find_workspace_root()
    metadata_file = workspace_root / f"bazel-bin/{package_path}/{target_name}_metadata.json"
    if not metadata_file.exists():
        return None

    with open(metadata_file) as f:
        return json.load(f)


def get_app_metadata(bazel_target: str) -> Dict:
    """Get release metadata for an app by building and reading its metadata target.
    
    Results are cached in-process to avoid redundant Bazel invocations.
    If the metadata file already exists on disk (e.g., from a batch build),
    it is read directly without invoking Bazel.
    
    Args:
        bazel_target: Full bazel target path (e.g., "//path/to/app:app_metadata")
    """
    if bazel_target in _metadata_cache:
        return _metadata_cache[bazel_target]

    # Try reading from disk first (may already be built by a batch build)
    metadata = _read_metadata_file(bazel_target)
    if metadata is None:
        # Build the metadata target
        run_bazel(["build", bazel_target])
        metadata = _read_metadata_file(bazel_target)
        if metadata is None:
            raise FileNotFoundError(f"Metadata file not found after building {bazel_target}")
    
    _metadata_cache[bazel_target] = metadata
    return metadata


def list_all_apps() -> List[Dict[str, str]]:
    """List all apps in the monorepo that have release metadata.
    
    Batch-builds all metadata targets in a single Bazel invocation to avoid
    repeated analysis overhead.
    
    Returns:
        List of dicts with full metadata for each app
    """
    # Query for all metadata targets
    result = run_bazel(["query", "kind(app_metadata, //...)", "--output=label"])

    targets = [line for line in result.stdout.strip().split('\n') if line and '_metadata' in line]
    
    if not targets:
        return []
    
    # Batch-build all metadata targets in a single Bazel invocation
    # This avoids N separate Bazel analysis phases
    run_bazel(["build"] + targets)

    apps = []
    for target in targets:
        try:
            # get_app_metadata will read from disk (already built) and cache the result
            metadata = get_app_metadata(target)
            metadata['bazel_target'] = target
            apps.append(metadata)
        except Exception as e:
            print(f"Warning: Could not get metadata for {target}: {e}")
            continue

    return sorted(apps, key=lambda x: x['name'])


def get_image_targets(bazel_target: str) -> Dict[str, str]:
    """Get all image-related targets for an app.
    
    Args:
        bazel_target: Full bazel target path (e.g., "//path/to/app:app_metadata")
    """
    # Extract package path from metadata target
    target_parts = bazel_target[2:].split(":")
    package_path = target_parts[0]
    
    # Get the metadata to find the actual image target name
    metadata = get_app_metadata(bazel_target)
    image_target_name = metadata['image_target']
    
    # Build full image target paths
    base_image_target = f"//{package_path}:{image_target_name}"
    
    return {
        "base": base_image_target,
        "push": f"{base_image_target}_push",
    }

File: tools/release/pkg/metadata/metadata.go
// Package metadata provides app metadata utilities for the release helper.
package metadata

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/whale-net/everything/tools/release/pkg/core"
)

// AppMetadata represents release metadata for an app.
type AppMetadata struct {
	Name         string            `json:"name"`
	AppType      string            `json:"app_type"`
	Version      string            `json:"version"`
	Description  string            `json:"description"`
	Registry     string            `json:"registry"`
	RepoName     string            `json:"repo_name"`
	ImageTarget  string            `json:"image_target"`
	Domain       string            `json:"domain"`
	Language     string            `json:"language"`
	Port         int               `json:"port,omitempty"`
	Replicas     int               `json:"replicas,omitempty"`
	Labels       map[string]string `json:"labels,omitempty"`
	Annotations  map[string]string `json:"annotations,omitempty"`
	Dependencies []string          `json:"dependencies,omitempty"`
}

// AppInfo represents basic app information.
type AppInfo struct {
	BazelTarget string `json:"bazel_target"`
	Name        string `json:"name"`
	Domain      string `json:"domain"`
}

// GetAppMetadata retrieves metadata for an app by building and reading its metadata target.
func GetAppMetadata(bazelTarget string) (*AppMetadata, error) {
	// Build the metadata target
	_, err := core.RunBazel([]string{"build", bazelTarget}, false, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to build metadata target: %w", err)
	}

	// Extract path from target for finding the generated file
	if !strings.HasPrefix(bazelTarget, "//") {
		return nil, fmt.Errorf("invalid bazel target format: %s", bazelTarget)
	}

	targetParts := strings.Split(bazelTarget[2:], ":")
	if len(targetParts) != 2 {
		return nil, fmt.Errorf("invalid bazel target format: %s", bazelTarget)
	}

	packagePath := targetParts[0]
	targetName := targetParts[1]

	// Read the generated JSON file
	workspaceRoot, err := core.FindWorkspaceRoot()
	if err != nil {
		return nil, fmt.Errorf("failed to find workspace root: %w", err)
	}

	metadataFile := filepath.Join(workspaceRoot, "bazel-bin", packagePath, targetName+"_metadata.json")
	data, err := os.ReadFile(metadataFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read metadata file %s: %w", metadataFile, err)
	}

	var metadata AppMetadata
	if err := json.Unmarshal(data, &metadata); err != nil {
		return nil, fmt.Errorf("failed to parse metadata JSON: %w", err)
	}

	return &metadata, nil
}

// ListAllApps lists all apps in the monorepo that have release metadata.
func ListAllApps() ([]AppInfo, error) {
	// Query for all metadata targets
	result, err := core.RunBazel([]string{"query", "kind(app_metadata, //...)", "--output=label"}, true, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to query app metadata: %w", err)
	}

	var apps []AppInfo
	for _, line := range result.Lines() {
		if !strings.Contains(line, "_metadata") {
			continue
		}

		// Get metadata to extract app name and domain
		metadata, err := GetAppMetadata(line)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Warning: Could not get metadata for %s: %v\n", line, err)
			continue
		}

		apps = append(apps, AppInfo{
			BazelTarget: line,
			Name:        metadata.Name,
			Domain:      metadata.Domain,
		})
	}

	// Sort by name
	sort.Slice(apps, func(i, j int) bool {
		return apps[i].Name < apps[j].Name
	})

	return apps, nil
}

// ImageTargets represents all image-related targets for an app.
type ImageTargets struct {
	Base      string
	AMD64     string
	ARM64     string
	PushBase  string
	PushAMD64 string
	PushARM64 string
}

// GetImageTargets returns all image-related targets for an app.
func GetImageTargets(bazelTarget string) (*ImageTargets, error) {
	// Extract package path from metadata target
	if !strings.HasPrefix(bazelTarget, "//") {
		return nil, fmt.Errorf("invalid bazel target format: %s", bazelTarget)
	}

	targetParts := strings.Split(bazelTarget[2:], ":")
	if len(targetParts) != 2 {
		return nil, fmt.Errorf("invalid bazel target format: %s", bazelTarget)
	}
	packagePath := targetParts[0]

	// Get the metadata to find the actual image target name
	metadata, err := GetAppMetadata(bazelTarget)
	if err != nil {
		return nil, fmt.Errorf("failed to get metadata: %w", err)
	}

	imageTargetName := metadata.ImageTarget

	// Build full image target paths
	baseImageTarget := fmt.Sprintf("//%s:%s", packagePath, imageTargetName)

	return &ImageTargets{
		Base:      baseImageTarget,
		AMD64:     baseImageTarget + "_amd64",
		ARM64:     baseImageTarget + "_arm64",
		PushBase:  baseImageTarget + "_push",
		PushAMD64: baseImageTarget + "_push_amd64",
		PushARM64: baseImageTarget + "_push_arm64",
	}, nil
}
